### PR TITLE
[Perf] DeepSeek-OCR-2 TTFT Optimize

### PR DESCRIPTION
**Pull request by LiuLi1998**
`create_custom_4d_mask` in `deepseek-ai/DeepSeek-OCR-2` was CPU-bound, 
截屏2026-07-23 13 54 44

## What changed

Optimize `create_custom_4d_mask` to computes the mask in one shot and caches it:

- Mask is batch-invariant (token_type_ids is the same pattern for every sample),
  so we compute one `[1, 1, S, S]` mask and broadcast over the batch.
- Cached by `(sequence_length, dtype, device)`; ≤2 tiny entries (<1 MB total).

## Benchmark

Hardware: NVIDIA A10. 
Model: `deepseek-ai/DeepSeek-OCR-2`. 
Workload: 100 requests, concurrency 10, 1 images, `max_tokens=1`
vLLM: 0.25.1
Server Setup: `vllm serve /root/models/DeepSeek-OCR-2 \
  --trust-remote-code \
  --max-model-len 4096 \
  --gpu-memory-utilization 0.85 \
  --mm-processor-cache-gb 0 \
  --no-enable-prefix-caching \
  --profiler-config.profiler=torch \
  --profiler-config.torch_profiler_dir=/root/vllm_trace \
  --port 8000`

| Metric | Before | After | Improvement |
|---|---|---|---|
| TTFT p50 | 4193 ms | 2283 ms | **1.84× / −46%** |
| TTFT p99 | 4207 ms | 2292 ms | 1.84× |
| TTFT mean | 4023 ms | 2198 ms | 1.83× |
| Throughput | 2.36 req/s | 4.31 req/s | **1.83×** |

torch profiler (same workload) confirms the mechanism:

| Op | Before | After |
|---|---|---|
| `aten::_index_put_impl_` calls | 229,296 | **0** |
| `aten::_index_put_impl_` CPU total | 52.2% (22.3 s) | 0 |
| `aten::copy_` calls | 532,122 | 71,592 |

Single request profiler after optimize: 
image

Benchmark Script:
```python
#!/usr/bin/env python3
"""TTFT benchmark for an OpenAI-compatible /v1/chat/completions endpoint.

Measures time-to-first-token (TTFT) via streaming, under configurable
concurrency. Supports text-only or image prompts, and can force cache-miss
by making each request's content unique.

Examples:
  # image (unique per request => cache miss), concurrency 8
  python ttft_bench.py --model /root/models/DeepSeek-OCR-2 \
      --n 100 --concurrency 10 --unique-image
"""
import argparse
import base64
import io
import json
import statistics
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests


def make_unique_image_b64(tag, w, h):
    from PIL import Image, ImageDraw

    im = Image.new("RGB", (w, h), "white")
    ImageDraw.Draw(im).text((30, 30), f"UNIQUE {tag}\nTTFT benchmark", fill="black")
    buf = io.BytesIO()
    im.save(buf, format="PNG")
    return base64.b64encode(buf.getvalue()).decode()


def build_payload(args, tag):
    is_image = args.image or args.unique_image
    if is_image:
        if args.unique_image:
            b64 = make_unique_image_b64(tag, args.img_w, args.img_h)
        else:
            with open(args.image, "rb") as f:
                b64 = base64.b64encode(f.read()).decode()
        content = [
            {"type": "text", "text": args.prompt},
            {"type": "image_url", "image_url": {"url": f"data:image/png;base64,{b64}"}},
        ]
    else:
        text = f"{args.prompt} [req {tag}]" if args.unique_text else args.prompt
        content = text
    return {
        "model": args.model,
        "messages": [{"role": "user", "content": content}],
        "max_tokens": args.max_tokens,
        "temperature": 0.0,
        "stream": True,
        "stream_options": {"include_usage": False},
    }


def one_request(args, tag):
    """Returns (ttft_ms, total_ms, n_chunks, error)."""
    payload = build_payload(args, tag)
    t0 = time.perf_counter()
    ttft = None
    n_chunks = 0
    try:
        r = requests.post(args.url, json=payload, stream=True, timeout=args.timeout)
        if r.status_code != 200:
            return None, None, 0, f"HTTP {r.status_code}: {r.text[:200]}"
        for line in r.iter_lines():
            if not line:
                continue
            s = line.decode("utf-8", "ignore")
            if not s.startswith("data: "):
                continue
            data = s[6:]
            if data == "[DONE]":
                break
            try:
                chunk = json.loads(data)
            except json.JSONDecodeError:
                continue
            choice = chunk["choices"][0]
            delta = choice.get("delta", {})
            has_content = bool(delta.get("content")) or bool(
                delta.get("reasoning_content")
            )
            if has_content:
                if ttft is None:
                    ttft = (time.perf_counter() - t0) * 1000
                n_chunks += 1
            if choice.get("finish_reason"):
                break
        total = (time.perf_counter() - t0) * 1000
        r.close()
    except Exception as e:  # noqa: BLE001
        return None, None, 0, repr(e)
    return ttft, total, n_chunks, None


def pct(vals, p):
    if not vals:
        return float("nan")
    k = min(len(vals) - 1, int(round(p / 100.0 * (len(vals) - 1))))
    return vals[k]


def summarize(name, vals):
    vals = sorted(v for v in vals if v is not None)
    if not vals:
        print(f"{name}: (no data)")
        return
    print(
        f"{name} ms:  min={vals[0]:.0f}  p50={pct(vals,50):.0f}  "
        f"p90={pct(vals,90):.0f}  p99={pct(vals,99):.0f}  max={vals[-1]:.0f}  "
        f"mean={statistics.mean(vals):.0f}"
    )


def run(args):
    for i in range(args.warmup):
        one_request(args, f"warmup{i}")

    ttfts, totals, errors = [], [], []
    t_start = time.perf_counter()
    with ThreadPoolExecutor(max_workers=args.concurrency) as ex:
        futs = [
            ex.submit(one_request, args, f"r{i}_{time.time_ns()}")
            for i in range(args.n)
        ]
        for fu in as_completed(futs):
            ttft, total, _, err = fu.result()
            if err:
                errors.append(err)
            else:
                ttfts.append(ttft)
                totals.append(total)
    wall = time.perf_counter() - t_start

    mode = "image" if (args.image or args.unique_image) else "text"
    print("\n=== TTFT benchmark ===")
    print(f"url={args.url}")
    print(f"model={args.model}  mode={mode}  max_tokens={args.max_tokens}")
    print(
        f"requests={args.n}  concurrency={args.concurrency}  "
        f"ok={len(ttfts)}  errors={len(errors)}"
    )
    print(f"wall={wall:.2f}s  throughput={len(ttfts)/wall:.2f} req/s")
    summarize("TTFT", ttfts)
    if args.max_tokens > 1:
        summarize("TOTAL", totals)
    if errors:
        print(f"first error: {errors[0]}")


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--url", default="http://127.0.0.1:8000/v1/chat/completions")
    ap.add_argument("--model", default="/root/models/DeepSeek-OCR-2")
    ap.add_argument("--n", type=int, default=20, help="total requests")
    ap.add_argument("--concurrency", type=int, default=1)
    ap.add_argument("--max-tokens", type=int, default=1,
                    help="1 => TTFT≈prefill; >1 also reports TOTAL latency")
    ap.add_argument("--prompt", default="Free OCR.")
    ap.add_argument("--warmup", type=int, default=2)
    ap.add_argument("--timeout", type=float, default=120)
    ap.add_argument("--image", default=None, help="fixed image file path (multimodal)")
    ap.add_argument("--unique-image", action="store_true",
                    help="generate a unique image per request (forces cache miss)")
    ap.add_argument("--img-w", type=int, default=1024)
    ap.add_argument("--img-h", type=int, default=768)
    ap.add_argument("--unique-text", action="store_true",
                    help="append a unique tag to the text prompt (forces cache miss)")
    run(ap.parse_args())


if __name__ == "__main__":
    main()
```

### CODE REVIEW

**Comment by LiuLi1998** [comment]
Hi @DarkLight1337 would you have time to take a look? Happy to address any feedback, thx

File: vllm/model_executor/models/deepencoder2.py [hunk 1]
```diff
@@ -158,36 +158,39 @@ def _create_custom_4d_mask(
...
+                            device=device,
+                        )
                     )
+                    allow = img[None, :] | (txt[:, None] & txt[None, :] & causal)
+                    base = torch.where(
+                        allow,
+                        torch.zeros((), dtype=dtype, device=device),
+                        torch.full((), min_dtype, dtype=dtype, device=device),
+                    )[None, None]
+                    cache[key] = base
```
Comment (Isotr0py) [inline vllm/model_executor/models/deepencoder2.py]: But `sequence_length` can be various, I think it will be inefficient for memory at low hit rate.

Comment (LiuLi1998) [reply vllm/model_executor/models/deepencoder2.py]: Thx for review, in practice this isn't a concern here, for two reasons:

1. The sequence_length only takes **two possible values** 288 or 512 in this path, so 
   the cache is bounded to at most 2 entries

2. Even **without** the cache, I measured a ~45% TTFT improvement from the 
   rest of the change. TTFT improvement mostly comes from **removing the CPU→GPU index and 
   memcpy** on this path, not from the cache itself.

Comment (Isotr0py) [reply vllm/model_executor/models/deepencoder2.py]: I see, if so, I think we just only need `sequence_length` as cache's key.

Comment (LiuLi1998) [reply vllm/model_executor/models/deepencoder2.py]: Makes sense. I've updated the code.

File: vllm/model_executor/models/deepencoder2.py [hunk 2]
```diff
@@ -158,36 +158,39 @@ def _create_custom_4d_mask(
...
```
Comment (Isotr0py) [inline vllm/model_executor/models/deepencoder2.py]: Perhaps we can make a cached method like this:
```python3
@cache
@classmethod
def compute_mask_base(cls, sequence_length: int, ...):
    ...
    return base
```
Then we can assign cache more elegant without `cache = self._mask_cache = {}`.

**Review by Isotr0py — APPROVED** [review]
LGTM, leave a nit.

**Comment by LiuLi1998** [comment]
@Isotr0py It seems that the `pre-run-check` is blocked. 
Could you help take a look at this?

File: vllm/model_executor/models/deepencoder2.py (hunk 2 above)
Comment (Isotr0py) [reply vllm/model_executor/models/deepencoder2.py]: @LiuLi1998 I just pushed https://github.com/vllm-project/vllm/pull/49531/commits/7ea00e7246222d9aed340567c72cf6dab50bb94c to implement this, which should work, can you help me double check this?

Comment (LiuLi1998) [reply vllm/model_executor/models/deepencoder2.py]: I think it's ok. I have tested it.
